src/examples/algorithms/Anomaly.cpp

```diff
@@ -0,0 +1,178 @@
+        std::vector<UInt>{COLS}  // Number of colums per dimension
+    };
+    // Temporal Pooler
+    Cells4 tp{
+        COLS,    // Number of columns
+        CELLS,     // Cells per column
+        12,
+        8,
+        15,
+        5,
+        .5,
+        .8,
+        1.0,
+        .1,
+        .1,
+        0.0,
+        false,
+        42,
+        true,
+        false //TODO use default params
+    };
+    // Anomaly
+    Anomaly anomaly{2, AnomalyMode::PURE, 0};
+
+    std::vector<UInt> lastTPOutput{COLS};
+
+  public:
+    Real compute(Real x) {
+        auto scalar = std::vector<Real32>(encoder.getOutputWidth());
+        encoder.encodeIntoArray(x, scalar.data());
+
+        // For some reason, the ScalarEncoder outputs a Real32 vector,
+        // which is incomaptible with the SpatialPooler's input, hence a new
+        // vector is created with the required type, and the elements
+        // are converted.
+        auto uint_scalar = VectorHelpers::castVectorType<Real32, UInt>(scalar);
+        if (DEBUG_LEVEL > 0) 
+        {
+          std::cout << x << ":\n";
+        } else if (DEBUG_LEVEL > 1) {
+          std::cout << "Scalar encoder: ";
+          VectorHelpers::print_vector(uint_scalar);
+        }
+
+        std::vector<UInt> spOutput(sp.getNumColumns());
+        sp.compute(uint_scalar.data(), true, spOutput.data());
+        sp.stripUnlearnedColumns(spOutput.data());
+        if (DEBUG_LEVEL > 2) {
+          std::cout << "Spatial pooler: ";
+          VectorHelpers::print_vector(spOutput);
+          std::cout << std::endl;
+        }
+
+        // The temporal pooler uses Real32, so the vector must be converted again
+        auto tpInput = VectorHelpers::castVectorType<UInt, Real>(spOutput);
+        std::vector<Real> tpOutput(tp.nCells());
+        tp.compute(tpInput.data(), tpOutput.data(), true, true);
+
+        // And the result is converted ONCE again to UInts for pretty printing to stdout.
+        auto uintTpOutput = VectorHelpers::castVectorType<Real32, UInt>(tpOutput);
+        if (DEBUG_LEVEL > 3) {
+          std::cout << "Output of temporal pooler: ";
+          VectorHelpers::print_vector(uintTpOutput);
+          std::cout << std::endl;
+        }
+
+        Real anScore = anomaly.compute(
+          VectorHelpers::binaryToSparse<UInt>(spOutput),     // Spatial pooler (current active columns)
+          VectorHelpers::binaryToSparse<UInt>(lastTPOutput), // Temporal pooler (previously predicted columns)
+          0,
+          0
+        );
+        // Save the output of the TP for the next iteration...
+        lastTPOutput = VectorHelpers::cellsToColumns(uintTpOutput, CELLS);
+
+        if(DEBUG_LEVEL > 0) {
+        std::cout << "Anomaly: "
+                  << anScore << "\n\n";
+        } else if (DEBUG_LEVEL > 4) {
+        std::cout << "Normalized TP Output: ";
+        VectorHelpers::print_vector(lastTPOutput);
+        }
+
+        return anScore;
+    }
+};
+
+int main()
```

Comment:
Why would you need another class? Just move this main function to `src/examples/hotgym.cpp`.

Basically, there are the reusable algorithm components that belong in `src/nupic/` but the example-specific code belongs in `src/examples/`.